core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/AspectModelHelper.java

```diff
@@ -0,0 +1,180 @@
+        aspectModel.getOperations().forEach(operation -> properties.addAll(getProperties(operation)));
+        final Set<Property> characteristicProperties = new HashSet<>();
+        properties
+                .forEach(property -> characteristicProperties.addAll(getProperties(property.getCharacteristic())));
+        properties.addAll(characteristicProperties);
+        return properties;
+    }
+
+    public static Set<Property> getProperties(final Operation operation) {
+        return Stream.concat(operation.getInput().stream(), operation.getOutput().stream())
+                .collect(Collectors.toSet());
+    }
+
+    public static Set<Constraint> getConstraints(final Property property) {
+        final Set<Constraint> constraints = new HashSet<>();
+        if (property.getCharacteristic() instanceof Trait) {
+            constraints.addAll(((Trait) property.getCharacteristic()).getConstraints());
+        }
+        return constraints;
+    }
+
+    @SuppressWarnings("squid:S2250")
+    //Amount of elements in list is regarding the amount of properties in aspect model. Even in bigger aspects this should not lead to performance issues
+    public static Set<Property> getProperties(final Characteristic characteristic) {
+        if (characteristicList.contains(characteristic)) {
+            return Set.of();
+        }
+        characteristicList.add(characteristic);
+        final Optional<Type> dataType = characteristic.getDataType();
+        if (dataType.isPresent() && !(dataType.get() instanceof Entity)) {
+            return Collections.emptySet();
+        }
+        final Set<Property> properties = dataType.stream()
+                .map(Entity.class::cast)
+                .map(HasProperties::getProperties)
+                .flatMap(Collection::stream)
+                .collect(Collectors.toSet());
+
+        final Set<Property> propertiesCharacteristics = new HashSet<>();
+        properties.forEach(
+                property -> propertiesCharacteristics.addAll(getProperties(property.getCharacteristic())));
+        properties.addAll(propertiesCharacteristics);
+        characteristicList.remove(characteristic);
+        return properties;
+    }
+
+    public static Entity resolveEntity(final SingleEntity singleEntity, final List<Entity> entities) {
+        return entities.stream()
+                .filter(entity -> singleEntity.getDataType()
+                        .map(Type::getUrn)
+                        .map(u -> u.equals(entity.getUrn())).orElse(false))
+                .findFirst()
+                .orElseThrow(() -> new DocumentGenerationException("Could not find entity " + singleEntity
+                        + " in list of entities: " + entities));
+    }
+
+    public static String getNameFromURN(final String urn) {
+        String[] parts = urn.split("#");
+        return parts.length == 2 ? parts[1] : urn;
+    }
+
+    @SuppressWarnings("squid:S3655")
```

Comment:
Please add a comment to the annotation that explains what is suppressed and why

Reply:
done